x/archivedb/db.go

```diff
@@ -0,0 +1,184 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package archivedb
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+
+	"github.com/ava-labs/avalanchego/database"
+)
+
+var (
+	_ database.Batch          = (*batchWithHeight)(nil)
+	_ database.KeyValueReader = (*dbHeightReader)(nil)
+)
+
+// ArchiveDb
+//
+// Creates a thin database layer on top of database.Database. ArchiveDb is an
+// append only database which stores all state changes happening at every block
+// height. Each record is stored in such way to perform both fast inserts and selects.
+//
+// Currently its API is quite simple, it has two main functions, one to create a
+// Batch write with a given block height, inside this batch entries can be added
+// with a given value or they can be deleted. It also provides a Get function
+// that takes a given key and a height.
+//
+//	The way it works is as follows:
+//		- Height: 10
+//			Set(foo, "foo's value is bar")
+//			Set(bar, "bar's value is bar")
+//		- Height: 100
+//			Set(foo, "updatedfoo's value is bar")
+//		- Height: 1000
+//			Set(bar, "updated bar's value is bar")
+//			Delete(foo)
+//
+// When requesting `Get(foo, 9)` it will return an errNotFound error because foo
+// was not defined at block height 9, it was defined later. When calling
+// `Get(foo, 99)` it will return a tuple `("foo's value is bar", 10)` returning
+// the value of `foo` at height 99 (which was set at height 10). If requesting
+// `Get(foo, 2000)` it will return an error because `foo` was deleted at height
+// 1000.
+type archiveDB struct {
+	ctx context.Context
+
+	currentHeight uint64
+
+	rawDB database.Database
+}
+
+var (
+	currentHeightKey = "archivedb.height"
+	ErrUnknownHeight = errors.New("unknown height")
+)
+
+type batchWithHeight struct {
+	db     *archiveDB
+	height uint64
+	batch  database.Batch
+}
+
+func NewArchiveDB(
+	ctx context.Context,
+	db database.Database,
+) (*archiveDB, error) {
+	var currentHeight uint64
+	height, err := db.Get([]byte(currentHeightKey))
+	if err != nil {
+		if !errors.Is(err, database.ErrNotFound) {
+			return nil, err
+		}
+		currentHeight = 0
+	} else {
+		currentHeight = binary.BigEndian.Uint64(height)
+	}
+
+	return &archiveDB{
+		ctx:           ctx,
+		currentHeight: currentHeight,
+		rawDB:         db,
+	}, nil
+}
+
+// Tiny wrapper on top Get() passing the last stored height
+func (db *archiveDB) GetLastBlock(key []byte) ([]byte, uint64, error) {
+	return db.Get(key, db.currentHeight)
+}
+
+// Returns an object which implements the database.KeyValueReader trait for all
```

Comment:
Could we update comments on exported functions to follow the convention of:
`// GetHeightReader returns ...`